File: tools/source-analyzers/BRD-static-analysis/src/brd_generator/modules/discovery.py
import os
from pathlib import Path
from typing import List, Dict, Set
from enum import Enum
from pydantic import BaseModel

class FileType(str, Enum):
    MODEL = "model"
    CONTROLLER = "controller"
    SERVICE = "service"
    UTILITY = "utility"
    CONFIG = "config"
    TEST = "test"
    UNKNOWN = "unknown"

class SourceFile(BaseModel):
    path: str
    name: str
    extension: str
    file_type: FileType
    language: str

class DiscoveryEngine:
    """
    Scans directories and classifies files based on heuristics.
    """
    def __init__(self, root_path: str, ignore_patterns: List[str] = None):
        self.root_path = Path(root_path)
        self.ignore_patterns = set(ignore_patterns or [])
        # Basic ignore list
        self.ignore_patterns.update({'.git', '__pycache__', 'node_modules', 'dist', 'build', 'venv', '.env'})

    def scan(self) -> List[SourceFile]:
        discovered_files = []
        
        for root, dirs, files in os.walk(self.root_path):
            # Prune ignored directories
            dirs[:] = [d for d in dirs if d not in self.ignore_patterns]
            
            for file in files:
                if file.startswith('.'): continue
                
                file_path = Path(root) / file
                if self._should_ignore(file_path):
                    continue
                
                source_file = self._classify_file(file_path)
                if source_file:
                    discovered_files.append(source_file)
                    
        return discovered_files

    def _should_ignore(self, path: Path) -> bool:
        # Simple containment check, could be enhanced with glob matching
        parts = set(path.parts)
        return not self.ignore_patterns.isdisjoint(parts)

    def _classify_file(self, path: Path) -> SourceFile:
        ext = path.suffix.lower()
        name = path.name.lower()
        parent = path.parent.name.lower()
        
        language = "unknown"
        if ext in ['.py']: language = "python"
        elif ext in ['.js', '.mjs', '.cjs']: language = "javascript"
        elif ext in ['.ts', '.tsx']: language = "typescript"
        elif ext in ['.rs']: language = "rust"
        else: return None # Skip unsupported files for now

        file_type = FileType.UNKNOWN
        
        # Heuristics
        if 'test' in name or 'spec' in name or 'tests' in path.parts:
            file_type = FileType.TEST
        elif 'model' in name or 'schema' in name or 'entity' in name or 'models' in path.parts:
            file_type = FileType.MODEL
        elif 'controller' in name or 'route' in name or 'api' in name or 'view' in path.parts:
            file_type = FileType.CONTROLLER
        elif 'service' in name or 'manager' in name:
            file_type = FileType.SERVICE
        elif 'util' in name or 'helper' in name or 'common' in path.parts:
            file_type = FileType.UTILITY
        elif 'config' in name or 'setting' in name:
            file_type = FileType.CONFIG

        return SourceFile(
            path=str(path),
            name=path.name,
            extension=ext,
            file_type=file_type,
            language=language
        )

File: tools/source-analyzers/BRD-static-analysis-javascript/src/modules/discovery.ts
import { globby } from 'globby';
import path from 'path';

export enum FileType {
    MODEL = "model",
    CONTROLLER = "controller",
    SERVICE = "service",
    UTILITY = "utility",
    CONFIG = "config",
    TEST = "test",
    UNKNOWN = "unknown"
}

export interface SourceFile {
    path: string;
    name: string;
    extension: string;
    fileType: FileType;
    language: string;
}

export class DiscoveryEngine {
    private rootPath: string;
    private ignorePatterns: string[];

    constructor(rootPath: string, ignorePatterns: string[] = []) {
        this.rootPath = path.resolve(rootPath);
        this.ignorePatterns = [
            ...ignorePatterns,
            '**/node_modules/**',
            '**/dist/**',
            '**/build/**',
            '**/coverage/**',
            '**/.git/**'
        ];
    }

    async scan(): Promise<SourceFile[]> {
        const paths = await globby(['**/*.{js,jsx,ts,tsx,mjs,cjs}'], {
            cwd: this.rootPath,
            ignore: this.ignorePatterns,
            absolute: true
        });

        return paths.map(filePath => this.classifyFile(filePath)).filter((f): f is SourceFile => f !== null);
    }

    private classifyFile(filePath: string): SourceFile | null {
        const ext = path.extname(filePath).toLowerCase();
        const name = path.basename(filePath).toLowerCase();
        const dir = path.dirname(filePath).toLowerCase();

        // Language
        let language = 'unknown';
        if (['.js', '.jsx', '.cjs', '.mjs'].includes(ext)) language = 'javascript';
        if (['.ts', '.tsx', '.mts', '.cts'].includes(ext)) language = 'typescript';

        // Heuristics
        let fileType = FileType.UNKNOWN;
        if (name.includes('test') || name.includes('spec') || dir.includes('__tests__')) {
            fileType = FileType.TEST;
        } else if (name.includes('model') || name.includes('schema') || name.includes('entity') || name.includes('type')) {
            fileType = FileType.MODEL;
        } else if (name.includes('controller') || name.includes('route') || name.includes('api')) {
            fileType = FileType.CONTROLLER;
        } else if (name.includes('service') || name.includes('manager')) {
            fileType = FileType.SERVICE;
        } else if (name.includes('util') || name.includes('helper') || name.includes('common')) {
            fileType = FileType.UTILITY;
        } else if (name.includes('config') || name.includes('setting')) {
            fileType = FileType.CONFIG;
        }

        return {
            path: filePath,
            name: path.basename(filePath),
            extension: ext,
            fileType,
            language
        };
    }
}
